src/unit.cpp

```diff
@@ -3,96 +3,140 @@
-Unit makeUnit(const boost::units::si::area &) { return {Unit::Id::Area}; }
-Unit makeUnit(const decltype(std::declval<boost::units::si::amount>() *
-                             std::declval<boost::units::si::amount>()) &) {
-  return {Unit::Id::CountsVariance};
+
+/// Get the Id corresponding to the underlying unit
+Unit::Id Unit::id() const {
+  // return Unit::Id::Dimensionless;
+  return std::visit(
+      overloaded{[](boost::units::si::dimensionless) {
+                   return Unit::Id::Dimensionless;
+                 },
+                 [](boost::units::si::length) { return Unit::Id::Length; },
+                 [](boost::units::si::area) { return Unit::Id::Area; },
+                 [](decltype(std::declval<boost::units::si::area>() *
+                             std::declval<boost::units::si::area>())) {
+                   return Unit::Id::AreaVariance;
+                 },
+                 [](neutron::tof::counts) { return Unit::Id::Counts; },
+                 [](decltype(std::declval<neutron::tof::counts>() *
+                             std::declval<neutron::tof::counts>())) {
+                   return Unit::Id::CountsVariance;
+                 },
+                 [](decltype(std::declval<boost::units::si::dimensionless>() /
+                             std::declval<boost::units::si::length>())) {
+                   return Unit::Id::InverseLength;
+                 },
+                 [](decltype(std::declval<boost::units::si::dimensionless>() /
+                             std::declval<boost::units::si::time>())) {
+                   return Unit::Id::InverseTime;
+                 },
+                 [](neutron::tof::energy) { return Unit::Id::Energy; },
+                 [](neutron::tof::wavelength) { return Unit::Id::Wavelength; },
+                 [](boost::units::si::time) { return Unit::Id::Time; },
+                 [](neutron::tof::tof) { return Unit::Id::Tof; },
+                 [](boost::units::si::mass) { return Unit::Id::Mass; },
+                 [](auto) -> Unit::Id {
+                   throw std::runtime_error("Unit not yet implemented");
+                 }},
+      m_unit);
 }
-Unit makeUnit(const decltype(std::declval<boost::units::si::dimensionless>() /
-                             std::declval<boost::units::si::length>()) &) {
-  return {Unit::Id::InverseLength};
+
+// Mutliplying two units together using std::visit to run through the contents
+// of the std::variant
+Unit operator*(const Unit &a, const Unit &b) {
+  return std::visit(
+      [](auto x, auto y) -> Unit::unit_t {
+        if constexpr (std::is_same<decltype(x),
+                                   boost::units::si::dimensionless>::value)
+          return {y};
+        else if constexpr (std::is_same<decltype(y),
+                                        boost::units::si::dimensionless>::value)
+          return {x};
+        else if constexpr (isVariantMember<decltype(x * y),
```

Comment:
This `isVariantMember` check is very nice @nvaytet! Since it is frequently applied would it be possible to create a helper `isKnownUnit` which could be used as:

`isKnownUnit<decltype(x * y)>::value` ?

Reply:
i didn't write that, Simon found it on stackoverflow: https://stackoverflow.com/questions/45892170/how-do-i-check-if-an-stdvariant-can-hold-a-certain-type

Reply:
Or just put it in a `constexpr` helper function, if it wokrs, it is even shorter:

```cpp
template <class T>
constexpr bool isKnownUnit(const T &) {
  return isVariantMember<T, unit_t>::value;
}

// elsewhere
if constexpr (isKnownUnit(m*s)) { ... }
```